torch/csrc/jit/tensorexpr/ir_simplifier.h

```diff
@@ -0,0 +1,542 @@
+namespace jit {
+namespace tensorexpr {
+
+// Uses the evaluator to fold an operation with constant terms.
+// Expr v must be evaluatable without Vars.
+static Expr* evaluateOp(const Expr* v) {
+  ExprHandle handle(v);
+  ExprEval<SimpleIREvaluator> eval(handle);
+
+  switch (v->dtype().scalar_type()) {
+#define TYPE_CASE(Type, Name)                                        \
+  case ScalarType::Name: {                                           \
+    Type val = eval.value<Type>();                                   \
+    return getImmediateByType(v->dtype().scalar_type(), val).node(); \
+  }
+    AT_FORALL_SCALAR_TYPES_AND(Half, TYPE_CASE);
+#undef TYPE_CASE
+    default:
+      LOG(FATAL) << "Unsupported datatype: " << v->dtype();
+      return nullptr;
+  }
+  return nullptr;
+} // namespace tensorexpr
+
+static const Expr* newBinaryOpOfType(
+    IRNodeType expr_type,
+    const Expr* lhs,
+    const Expr* rhs,
+    bool option) {
+  switch (expr_type) {
+    case IRNodeType::kAdd:
+      return new Add(lhs, rhs);
+    case IRNodeType::kSub:
+      return new Sub(lhs, rhs);
+    case IRNodeType::kMul:
+      return new Mul(lhs, rhs);
+    case IRNodeType::kDiv:
+      return new Div(lhs, rhs);
+    case IRNodeType::kMod:
+      return new Mod(lhs, rhs);
+    case IRNodeType::kMax:
+      return new Max(lhs, rhs, option);
+    case IRNodeType::kMin:
+      return new Min(lhs, rhs, option);
+    case IRNodeType::kAnd:
+      return new And(lhs, rhs);
+    case IRNodeType::kXor:
+      return new Xor(lhs, rhs);
+    case IRNodeType::kLshift:
+      return new Lshift(lhs, rhs);
+    case IRNodeType::kRshift:
+      return new Rshift(lhs, rhs);
+    default:
+      LOG(FATAL) << "unsupported expr_type: " << static_cast<int>(expr_type);
+      return nullptr;
+  }
+}
+
+/* Interprets expr as an Immediate and returns the value as type T. */
+template <typename T>
+T immediateAs(const Expr* expr) {
+  T val{0};
+  switch (expr->dtype().scalar_type()) {
+#define TYPE_CASE(Type, Name)                                          \
+  case ScalarType::Name:                                               \
+    if (const Name##Imm* imm = dynamic_cast<const Name##Imm*>(expr)) { \
+      val = imm->value();                                              \
+    } else {                                                           \
+      LOG(FATAL) << "Bad expr: " << *expr << "\n";                     \
+    }                                                                  \
+    break;
+    AT_FORALL_SCALAR_TYPES_AND2(Half, Bool, TYPE_CASE);
+#undef TYPE_CASE
+    default:
+      LOG(FATAL) << "Unsupported datatype: " << expr->dtype();
+  }
+
+  return val;
+}
+
+/* Takes a LinearForm and converts it to Mul + (Add/Sub). */
+const Expr* expandLinearForm(const LinearForm* v, IRMutator* mutator) {
+  const Expr* mul = nullptr;
+  const Expr* A = v->getA();
+  const Expr* B = v->getB();
+  // X can be a LinearForm.
+  const Expr* X = v->getX()->accept_mutator(mutator);
```

Comment:
This part seems to be unnecessarily bundled here. Why do we need to run mutator on X as we expanding? Can it be run before `expandLinearForm` where it is called?

Reply:
Yeah, maybe. I'll take a look. It's a bit complex because this method is used by two different mutators.

Reply:
Ok so the issue here is that if (say, in the TermExpander) we recurse the tree first we still need a linear term to pass into this func. We'll end up creating a bunch of temporary objects we don't need.

Could this be addressed via comments?

Reply:
Yeah, if there is a justification, I'm fine with that. But it seems that in the new version of your PR `mutator` is unused at all?

Reply:
Uhhh guess we don't need it then? I was pretty sure I had a test that needed it but I've triple checked and everything is fine. Cool.